### ISSUE TITLE
[question] multicast vs unicast

### ISSUE THREAD

**Issue by Karsten1987** [eProsima/Fast-RTPS#63]
I ran into a couple of for me unclear things and wanted to sort them out here.
All I encountered is based on [this](https://github.com/ros2/demos/tree/master/image_tools) ROS2 demo.

**1. Disabled all network devices except loopback adapter.**  

I run all publisher/subscriber only locally on my machine, using the loopback device `lo` for communicating. 
When doing so, the demo application aborts directly with the following message

```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >'
  what():  set_option: No such device
Aborted
```

From what I understand is that there is no multicast enabled device available. Does that mean, Fast-RTPS does necessarily require a multicast enabled network interface?

**2. Enabling multicast on loopback adapter**

Given the error in 1.) I explicitly enabled the multicast functionality on the `lo`. 

```
sudo ifconfig lo mulitcast
sudo route add -net 224.0.0.0 netmask 240.0.0.0 dev lo
```

With this, the demo application successfully runs and the publisher/subscriber communication works over the `lo`. 

Output from wireshark:

```
Internet Protocol Version 4, Src: 0.0.0.0, Dst: 239.255.1.4
[...]
Time to live: 1
[...]
Default port mapping: MULTICAST_USERTRAFFIC, domainId=71
[...]
fragementSize: 64964
```

**3. Enabling wireless adapter**
With both adapter enabled (`lo` and `wlan0`) I run the same demo again. The communication works just fine, however in wireshark I see RTPS communication on both devices - one multicast, one unicast.

Output from wireshark (1):

```
Internet Protocol Version 4, Src: 10.0.0.177, Dst: 239.255.1.4
[...]
Time to live: 1
[...]
Default port mapping: MULTICAST_USERTRAFFIC, domainId=71
[...]
fragementSize: 64964
```

Output from wireshark (2):

```
Internet Protocol Version 4, Src: 10.0.0.177, Dst: 10.0.0177
[...]
Time to live: 64
[...]
Default port mapping: domainId=71, participantIdx=0, nature=UNICAST_USERTRAFFIC
[...]
fragementSize: 64964
```

The actual question here now is how Fast-RTPS decides which device is going to be taken for communication and why there is traffic on both devices in 3.) with one sending data in multicast and the other in unicast.

**Comment by edwinferretjp** [comment]
**Question**
This might not make much sense since I appreciate Fast-RTPS is designed for communication over a network but is there, by any chance, a setting that can be enabled at runtime/compile-time which does not require a network device/loopback for participant discovery and communication? This would be useful for the case where it is known in advance that all publishers and subscribers are running in the same process and that there are no other publishers or subscribers outside of this process, or even if there are, that they should not be communicated with.

**Context**
I'm asking this in relation to my question on ROS2 (https://github.com/ros2/demos/issues/128) where even when the publishers and subscribers run in the same process, meta-data is still sent through the underlying DDS layer, even though the message data can short-circuit the DDS layer and be copied in-memory. 

**Why?**
It would be nice(for me at least!) to have the option of not requiring a network connection for embedded applications on devices that have no network adaptor, or where at run-time the application discovers it is impossible to communicate (even with itself) over the available network device (e.g. an iOS/Android phone connected only to a 3G/4G telephony network). 

Having this option in Fast-RTPS would allow applications that do have a network connection to exploit the full functionality of network-connected applications while not completely hobbling all functionality of an application when no network is available. In both cases, I want to keep the Fast-RTPS layer even if there is no network available.

I appreciate this is probably not the intended use of Fast-RTPS but it is nonetheless a practical consideration that might occur often in reality.

**Comment by ufechner7** [comment]
This is annoying, even though the workaround to enable multicasting on the loop-back adapter works for me. I would like to fix this. At which source code file should I look?

**Comment by edwinferretjp** [comment]
@ufechner7 The loop-back works for me on Mac and I'm sure would work on Linux though I haven't tried. Do you have a loopback solution that works on Android and/or iOS devices too by any chance?

**Comment by MiguelCompany** [comment]
This issue should have been fixed by #303, as it specifically handled the case where all network interfaces except loopback adapter are disabled.
